### PR TITLE
Add OS X support for Sockets and address CR feedback from #10.

### PR DESCRIPTION
**Pull request by pgavlin**

### CODE REVIEW

**Comment by pgavlin** [comment]
@CIPop @stephentoub PTAL

File: src/System.Net.Sockets/src/System.Net.Sockets.csproj [hunk 1]
```diff
@@ -9,8 +9,6 @@
     <DefineConstants>$(DefineConstants);FEATURE_CORECLR</DefineConstants>
   </PropertyGroup>
   <!-- Help VS understand available configurations -->
-  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'FreeBSD_Debug|AnyCPU' " />
```
Comment (CIPop) [inline src/System.Net.Sockets/src/System.Net.Sockets.csproj:12]: Why are we removing FreeBSD? I thought this is community supported.

**Comment by CIPop** [comment]
:+1: Reviewed common and Windows code. A quick look at the non-Windows code didn't reveal anything.

File: src/System.Net.Sockets/src/System.Net.Sockets.csproj (hunk 1 above)
Comment (pgavlin) [inline src/System.Net.Sockets/src/System.Net.Sockets.csproj:12]: I'm pretty sure that we shouldn't have configurations that don't build in our project files. This configuration can be re-added when the community adds support for FreeBSD to S.N.Sockets.

**Comment by pgavlin** [comment]
@stephentoub do you think you'll have a chance to look at this in the near future?

**Comment by stephentoub** [comment]
Sorry for the delay, Pat.  I'll review it tonight.  If you like, feel free to go ahead with merging, and if I have any feedback, you can address it as is appropriate subsequently.

**Comment by pgavlin** [comment]
No worries--I'll go ahead and merge this now, as I have some dependent changes that I'd like to send out.

File: src/Common/src/Interop/OSX/libc/Interop.SocketOptionName.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+internal static partial class Interop
+{
+    // NOTE: this is only correct for Linux on x86-64
```
Comment (stephentoub) [inline src/Common/src/Interop/OSX/libc/Interop.SocketOptionName.cs:6]: This comment is inaccurate, right?

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 3]
```diff
@@ -157,35 +147,42 @@ public void Dequeue()
...
+        private object _queueLock = new object();
 
         public SocketAsyncContext(int fileDescriptor, SocketAsyncEngine engine)
         {
             _fileDescriptor = fileDescriptor;
             _engine = engine;
+            _handle = GCHandle.Alloc(this, GCHandleType.Normal);
 
-            var events = Interop.libc.EPOLLIN | Interop.libc.EPOLLOUT | Interop.libc.EPOLLRDHUP | Interop.libc.EPOLLET;
+            var events = SocketAsyncEvents.Read | SocketAsyncEvents.Write;
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:178]: Could `events` here be a const?

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 4]
```diff
@@ -247,7 +246,11 @@ public SocketError AcceptAsync(byte[] socketAddress, int socketAddressLen, Actio
             SocketError errorCode;
             if (TryCompleteAccept(_fileDescriptor, socketAddress, ref socketAddressLen, out acceptedFd, out errorCode))
             {
-                ThreadPool.QueueUserWorkItem(_ => callback(acceptedFd, socketAddress, socketAddressLen, errorCode), null);
+                ThreadPool.QueueUserWorkItem(args =>
+                {
+                    var tup = (Tuple<int, byte[], int, SocketError>)args;
+                    callback(tup.Item1, tup.Item2, tup.Item3, tup.Item4);
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:252]: `callback` here should be passed in the Tuple as well.  As it is, both the closure and the delegate are still getting allocated to pass callback in.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 5]
```diff
@@ -320,7 +323,7 @@ public SocketError ConnectAsync(byte[] socketAddress, int socketAddressLen, Acti
             SocketError errorCode;
             if (TryCompleteConnect(_fileDescriptor, socketAddress, socketAddressLen, out errorCode))
             {
-                ThreadPool.QueueUserWorkItem(_ => callback(errorCode), null);
+                ThreadPool.QueueUserWorkItem(arg => callback((SocketError)arg), errorCode);
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:326]: I'm assuming we'll go back through later and clean up these closures.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 6]
```diff
@@ -538,7 +525,11 @@ public SocketError ReceiveFromAsync(byte[] buffer, int offset, int count, int fl
             SocketError errorCode;
             if (TryCompleteReceiveFrom(_fileDescriptor, buffer, offset, count, flags, socketAddress, ref socketAddressLen, out bytesReceived, out receivedFlags, out errorCode))
             {
-                ThreadPool.QueueUserWorkItem(_ => callback(bytesReceived, socketAddress, socketAddressLen, receivedFlags, errorCode), null);
+                ThreadPool.QueueUserWorkItem(args =>
+                {
+                    var tup = (Tuple<int, byte[], int, int, SocketError>)args;
+                    callback(tup.Item1, tup.Item2, tup.Item3, tup.Item4, tup.Item5);
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:531]: Same comment as earlier about the `callback` capture.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 7]
```diff
@@ -586,7 +577,11 @@ public SocketError ReceiveFromAsync(IList<ArraySegment<byte>> buffers, int flags
             SocketError errorCode;
             if (TryCompleteReceiveFrom(_fileDescriptor, buffers, flags, socketAddress, ref socketAddressLen, out bytesReceived, out receivedFlags, out errorCode))
             {
-                ThreadPool.QueueUserWorkItem(_ => callback(bytesReceived, socketAddress, socketAddressLen, receivedFlags, errorCode), null);
+                ThreadPool.QueueUserWorkItem(args =>
+                {
+                    var tup = (Tuple<int, byte[], int, int, SocketError>)args;
+                    callback(tup.Item1, tup.Item2, tup.Item3, tup.Item4, tup.Item5);
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:583]: Ditto.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 8]
```diff
@@ -988,38 +987,36 @@ public unsafe void HandleEvents(uint events)
...
                         acceptOrConnectQueue = _acceptOrConnectQueue.Stop();
                         sendQueue = _sendQueue.Stop();
                         receiveQueue = _receiveQueue.Stop();
 
-                        if (_registeredEvents != 0)
+                        if (_registeredEvents != SocketAsyncEvents.None)
                         {
                             Interop.Error errorCode;
-                            if (!_engine.TryUnregister(ref _handle, _fileDescriptor, 0, out errorCode))
+                            if (!_engine.TryRegister(_fileDescriptor, _registeredEvents, SocketAsyncEvents.None, _handle, out errorCode))
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:1006]: The comment above says "Drain queues and unregister events", but here we're calling TryRegister.  Are both correct?

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+                (((events & SocketAsyncEvents.Close) != 0) ? Interop.libc.EPOLLHUP : 0) |
+                (((events & SocketAsyncEvents.Error) != 0) ? Interop.libc.EPOLLERR : 0);
+        }
+
+        public unsafe void EventLoop()
+        {
+            var events = stackalloc Interop.libc.epoll_event[64];
+            for (;;)
+            {
+                int numEvents = Interop.libc.epoll_wait(_epollFd, events, 64, -1);
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs:58]: Consider moving the `64` into a named constant that can be used both here and in the stackalloc.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs (hunk 8 above)
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:1006]: Nevermind, I see now that TryRegister handles both add and delete based on the args supplied to it.

**Comment by stephentoub** [comment]
Looks good, Pat. Thanks.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs (hunk 9 above)
Comment (pgavlin) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs:58]: Done.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs (hunk 4 above)
Comment (pgavlin) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:252]: Done.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs (hunk 3 above)
Comment (pgavlin) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:178]: Done.

File: src/Common/src/Interop/OSX/libc/Interop.SocketOptionName.cs (hunk 2 above)
Comment (pgavlin) [inline src/Common/src/Interop/OSX/libc/Interop.SocketOptionName.cs:6]: Yes; fixed.

**Pull request merged**